Kernel/VirtIO/VirtIO.h

```diff
@@ -0,0 +1,251 @@
+        }
+
+        template<typename T>
+        void write(u32 offset, T value)
+        {
+            if (!base)
+                return;
+            ASSERT(size >= sizeof(T));
+            ASSERT(offset + sizeof(T) <= size);
+            *(volatile T*)(base->vaddr().offset(offset).get()) = value;
+        }
+    };
+
+    struct Configuration {
+        u8 cfg_type;
+        u8 bar;
+        u32 offset;
+        u32 length;
+    };
+
+    const Configuration* get_config(u8 cfg_type, u32 index = 0) const
+    {
+        for (const auto& cfg : m_configs) {
+            if (cfg.cfg_type != cfg_type)
+                continue;
+            if (index > 0) {
+                index--;
+                continue;
+            }
+            return &cfg;
+        }
+        return nullptr;
+    }
+    const Configuration* get_common_config(u32 index = 0) const;
+    const Configuration* get_device_config(u32 index = 0) const;
+
+    template<typename F>
+    void read_config_atomic(F f)
+    {
+        if (m_common_cfg) {
+            u8 generation_before, generation_after;
+            do {
+                generation_before = config_read8(m_common_cfg, 0x15);
+                f();
+                generation_after = config_read8(m_common_cfg, 0x15);
+            } while (generation_before != generation_after);
+        } else {
+            f();
+        }
+    }
+
+    u8 config_read8(const Configuration*, u32);
+    u16 config_read16(const Configuration*, u32);
+    u32 config_read32(const Configuration*, u32);
+    void config_write8(const Configuration*, u32, u8);
+    void config_write16(const Configuration*, u32, u16);
+    void config_write32(const Configuration*, u32, u32);
+    void config_write64(const Configuration*, u32, u64);
+
+    auto mapping_for_bar(u8) -> MappedMMIO&;
+
+    u8 read_status_bits();
+    void clear_status_bit(u8);
+    void set_status_bit(u8);
+    u64 get_device_features();
+    bool finish_init();
+
+    VirtIOQueue* get_queue(u16 queue_index)
+    {
+        return &m_queues[queue_index];
+    }
+    void set_requested_queue_count(u16);
+
+    template<typename F>
+    bool negotiate_features(F f)
+    {
+        u64 device_features = get_device_features();
+        u64 accept_features = f(device_features);
+        ASSERT(!(~device_features & accept_features));
+        return accept_device_features(device_features, accept_features);
+    }
+
+    static bool is_feature_set(u64 feature_set, u64 test_feature)
+    {
+        // features can have more than one bit
+        return (feature_set & test_feature) == test_feature;
+    }
+    bool is_feature_accepted(u64 feature) const
+    {
+        ASSERT(m_did_accept_features);
+        return is_feature_set(m_accepted_features, feature);
+    }
+
+    void supply_buffer_and_notify(u16 queue_index, const u8* buffer, u32 len, BufferType);
+
+    virtual void handle_irq(const RegisterState&) override;
+    virtual void handle_device_config_change() = 0;
```

Comment:
This method doesn't return any result to the user if it succeeded or not. Is there a mechanism inside such method to set a flag in case of failure?